csharp/ql/test/library-tests/csharp11/cil/Struct.cs_

```diff
@@ -0,0 +1,11 @@
+﻿namespace structassembly;
+
+public class MyEmptyClass { }
+
+public ref struct RefStruct
+{
+    public int MyInt;
+    public ref byte MyByte;
+    public ref object MyObject;
+    internal ref MyEmptyClass MyEmptyClass;
```

Comment:
Sure! I will do that!

Reply:
I checked it and you are right. The cil type annotation implementation reaches it's limit with these examples.
`ref readonly` means that the field is a reference field that refers to a readonly value. This behaves as expected in terms of extraction.
`readonly ref` means that the field is read(init)only of a reference type. In this case the reference keyword belongs to the type and not the field.
If we wan't to cater for this semantic difference it will require a change similar to the one you suggest in your comments (that the idea of cil_type_annotations) needs to be changed.
The question is, whether this slight discrepancy should block further implementation? (at the very least we need to record this as an issue).

Reply:
I'm not so sure there's such an obvious differentiation between `ref` being a field modifier, or `ref` being a type modifier if you look at the source code. Our current implementation is `ref` being a field modifier. This is conceptually matching the `ref` parameter modifier in `void M(ref int x)`. 

At the same time, in IL, the type of the parameter is `int&`, and the same is true for a `ref` field. So if you consider the meaning of the modifier, then my opinion is that `ref` is a modifier that changes the type.

Finally, to me `readonly` seems to be simply a modifier which exists in source code.

Reply:
After the offline discussion, I agree with you.
I will create an issue in the C# repo to start a discussion around the need for changing this. The incremental introduction of `ref` in the C# language, might mean, that we need to change our perception and implementation of `ref`.